File: server/routes/translate.ts
/**
 * Server-side translation proxy route — **ported away** (pinakes:64 US-1,
 * docs/UNIFIED-PROJECT-PLAN.md §7).
 *
 * `POST /api/translate` is served by the Python service now
 * (`services/api/src/pinakes/routers/translate.py` over `pinakes.ingest.translate`).
 * The handler here answers **501**.
 *
 * **The security posture is unchanged and is the whole point of the route**
 * (US-002, docs/SECURITY.md): `GOOGLE_TRANSLATE_API_KEY` is read server-side and
 * the upstream call is made server-side, so the browser never holds the key.
 * That is now true of the *Python* server; `server/security/translate-proxy.test.ts`
 * still guards the invariant that no `web/` source names the key, and it does not
 * care which backend answers. The three status codes the client depends on are
 * reproduced exactly — 400 bad body, **503 when no key is configured** (which is
 * what makes translation an optional enhancement `web/src/lib/scraping.ts` falls
 * through), 502 upstream failure.
 *
 * **`server/services/translate.ts` is NOT retired.** It is the graded spec for
 * the port; `services/api/tests/test_translate.py` is its suite, case for case.
 *
 * The path stays registered — the parity baseline was harvested from that path
 * set (see `routes/text-extractor.ts`).
 */

import { type Express, type Request, type Response } from "express";

/** The Python module that now serves this route. */
export const PORTED_TO = "services/api/src/pinakes/routers/translate.py";

/** The route this backend handed over. */
export const PORTED_ROUTES = {
  post: ["/api/translate"],
} as const;

/** Machine-readable discriminator in a retired route's body. */
export const PORTED_ERROR = "ported";

/** A handler for a route this backend no longer owns. 501, not 404 or 503. */
function portedToPython(route: string) {
  return (_req: Request, res: Response): void => {
    res.status(501).json({
      error: PORTED_ERROR,
      message:
        `${route} has been ported to the Python service and is served there ` +
        `(${PORTED_TO}). The Express handler is retired.`,
      route,
      servedBy: PORTED_TO,
      coverage: "/api/_parity/coverage",
    });
  };
}

export function registerTranslateRoutes(app: Express): void {
  for (const route of PORTED_ROUTES.post) {
    app.post(route, portedToPython(`POST ${route}`));
  }
}

File: services/api/src/pinakes/routers/translate.py
"""The `POST /api/translate` route — the server-side translation proxy.

Ported off `server/routes/translate.ts` (pinakes:64 US-1) over
:mod:`pinakes.ingest.translate`, which holds the key handling and the upstream
call. This file is the four status codes.

The 503 is the one worth knowing: with no ``$GOOGLE_TRANSLATE_API_KEY`` the
route answers "translation is not available" and `web/src/lib/scraping.ts`
silently falls through to its next source. That is the **default** state of a
checkout, not a misconfiguration — the same optional-enhancement shape as
``$GEONAMES_USERNAME`` in the place resolver.

Sync, not `async def`: the upstream call blocks (see
:mod:`pinakes.ingest.http`).
"""

from __future__ import annotations

import logging
from typing import Annotated, Any

from fastapi import APIRouter, Body
from fastapi.responses import JSONResponse

from pinakes.ingest import translate

logger = logging.getLogger("pinakes.translate")

router = APIRouter(tags=["translate"])

#: The 502 body. A failure upstream is never described to the client in detail —
#: the message could carry the key back out (`docs/SECURITY.md`).
FAILED = "Translation failed"


@router.post("/api/translate")
def translate_route(body: Annotated[Any, Body()] = None) -> Any:
    """Translate one string with the server-side key. 200/400/502/503."""
    try:
        request = translate.validate_translate_input(body)
    except translate.TranslateValidationError as error:
        return JSONResponse(status_code=400, content={"message": str(error)})

    try:
        return translate.translate_text(request)
    except translate.TranslateNotConfiguredError:
        return JSONResponse(
            status_code=503,
            content={
                "message": (
                    "Translation is not available (no server-side key configured)"
                )
            },
        )
    except translate.TranslateError:
        return JSONResponse(status_code=502, content={"message": FAILED})
    except Exception:  # noqa: BLE001 - the Express catch-all, same 502
        logger.exception("Translation failed")
        return JSONResponse(status_code=502, content={"message": FAILED})
